src/NuGet.Core/NuGet.Build.Tasks/RestoreTaskEx.cs

```diff
@@ -253,33 +253,35 @@ internal string GetProcessFileName(string processFileName)
 #endif
         }
 
-        private Dictionary<string, string> GetGlobalProperties()
+        /// <summary>
+        /// Enumerates a list of global properties for the current MSBuild instance.
+        /// </summary>
+        /// <returns>An <see cref="IEnumerable{T}" /> of <see cref="KeyValuePair{TKey, TValue}" /> objects containing global properties.</returns>
+        internal IEnumerable<KeyValuePair<string, string>> GetGlobalProperties()
         {
+            IReadOnlyDictionary<string, string> globalProperties = null;
+
 #if IS_CORECLR
             // MSBuild 16.5 and above has a method to get the global properties, older versions do not
-            Dictionary<string, string> msBuildGlobalProperties = BuildEngine is IBuildEngine6 buildEngine6
-                ? buildEngine6.GetGlobalProperties().ToDictionary(i => i.Key, i => i.Value, StringComparer.OrdinalIgnoreCase)
-                : new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            if (BuildEngine is IBuildEngine6 buildEngine6)
+            {
+                globalProperties = buildEngine6.GetGlobalProperties();
+            }
 #else
-            var msBuildGlobalProperties = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
-
             // MSBuild 16.5 added a new interface, IBuildEngine6, which has a GetGlobalProperties() method.  However, we compile against
             // Microsoft.Build.Framework version 4.0 when targeting .NET Framework, so reflection is required since type checking
             // can't be done at compile time
-            var buildEngine6Type = typeof(IBuildEngine).Assembly.GetType("Microsoft.Build.Framework.IBuildEngine6");
+            Type buildEngine6Type = typeof(IBuildEngine).Assembly.GetType("Microsoft.Build.Framework.IBuildEngine6");
```

Comment:
Why we're always doing this step? Considering we may already have populated `globalProperties` in line 268, I believe we do line#274-284 if above is failed to get `globalProperties` above, otherwise it's perf penalty. Because reflection is slow.

Reply:
This is because for .NET Framework we compile against MSBuild version 4.0.0.0 which does not have the interface `IBuildEngine6`.  For .NET Core we build against the MSBuild NuGet packages which have this interface.  If we ever have our .NET Framework assemblies target a newer MSBuild, we can get rid of this reflection code

Reply:
In that case shouldn't we do in reverse order?
If value is set with .NET Framework then don't do the other one then we don't overwrite value.

Reply:
Its confusing but line 268 is in an `#if IS_CORECLR` so that code is not compiled into the .NET Framework version of the assembly.

Reply:
I see. Do we generate .NetCore version of the assembly?

Reply:
Yes we ship the .NET Core assembly in the .NET SDK and the .NET Framework assembly in Visual Studio